### PR TITLE
GPT tagging for charts

### PR DESCRIPTION
**Pull request by mlbrgl**
Demo at http://staging-site-ai-tag-charts/admin/charts

- let GPT select topic tags based on chart metadata (title, subtitle, origin URL)
- allow for approval workflow _a posteriori_: tags associations are saved but marked for ulterior review
- key chart levels only available once a tagged has been approved
- uses the slightly faster `gpt-3.5-turbo` but we might upgrade to `gpt-4` if accuracy is better (not obvious so far)

- [ ] (deploy) add `OPENAI_API_KEY` key to .env

-> see [slack announcement](https://owid.slack.com/archives/C014SGT6W9X/p1694436797261319) for more details and demo

### CODE REVIEW

**Review by ikesau — APPROVED** [review]
Okay well this took me a while to troubleshoot, but once I got my environment working, it worked nicely!

I noticed a couple of strange results if you click "Suggest" multiple times (Adding "Global education" to a chart about fishing (maybe because of "[Schools of fish](https://en.wikipedia.org/wiki/Shoaling_and_schooling)" ? 😅)

I have a hunch that we may want the number of suggestions to be a bit more flexible (2-4 results) but we'll see how we go.

Authors may also want a slightly faster way to remove unapproved suggestions (a little ❌ on the chip) but maybe it's easiest to keep that in the edit menu. Up to you 🙂

File: adminSiteServer/apiRouter.ts [hunk 1]
```diff
@@ -2590,12 +2600,77 @@ apiRouter.post(
...
+        const topics: TagReactTagAutocomplete[] = await db.queryMysql(`
+        SELECT t.id, t.name
+        FROM tags t 
+        WHERE t.isTopic IS TRUE
+        AND t.parentId IN (${PUBLIC_TAG_PARENT_IDS.join(",")})
+    `)
+
+        const prompt = `
+            You will be provided with the chart metadata (delimited with XML tags),
+            as well as a list of possible topics (delimited with XML tags).
```
[the remark this answers is not in the document]
Comment (mlbrgl) [reply adminSiteServer/apiRouter.ts:2633]: Using clear delimiters is one of the recommended tactic to write clear instructions: https://platform.openai.com/docs/guides/gpt-best-practices/tactic-use-delimiters-to-clearly-indicate-distinct-parts-of-the-input. XML tags are just one of the ways to do that

File: adminSiteServer/apiRouter.ts [hunk 2]
```diff
@@ -2590,12 +2600,77 @@ apiRouter.post(
...
         gdoc.tags = tags
         await gdoc.save()
         return { success: true }
     }
 )
 
+apiRouter.get(
+    `/gpt/suggest-topics/${TaggableType.Charts}/:chartId.json`,
+    async (req: Request, res: Response): Promise<Record<"topics", Tag[]>> => {
+        const openai = new OpenAI()
```
[the remark this answers is not in the document]
Comment (mlbrgl) [reply adminSiteServer/apiRouter.ts:2614]: yep thanks, I've made this [more explicit in the next branch](https://github.com/owid/owid-grapher/pull/2636/commits/a42c66b9ec12d7dedafe4a69e9e9d0baac06ac61)

File: adminSiteServer/apiRouter.ts [hunk 3]
```diff
@@ -2590,12 +2600,77 @@ apiRouter.post(
...
+        if (!chart)
+            throw new JsonError(
+                `No chart found for id ${req.params.chartId}`,
+                404
+            )
+
+        const topics: TagReactTagAutocomplete[] = await db.queryMysql(`
+        SELECT t.id, t.name
+        FROM tags t 
+        WHERE t.isTopic IS TRUE
```
[the remark this answers is not in the document]
Comment (mlbrgl) [reply adminSiteServer/apiRouter.ts:2627]: ah thanks for catching this. Let's be complete here, even though you're right, we probably won't run into this edge case once everyone'DB has been updated with a recent prod dump.

**Comment by mlbrgl** [comment]
Thanks for the review @ikesau!

> I noticed a couple of strange results if you click "Suggest" multiple times 

Yes it does get quite creative sometimes 😅
 
> I have a hunch that we may want the number of suggestions to be a bit more flexible (2-4 results) but we'll see how we go.

Possibly, yes. We thought we probably want more than 1 but were concerned that 3 might start bringing too many low-quality matches. Let's see how this goes.

> Authors may also want a slightly faster way to remove unapproved suggestions (a little ❌ on the chip) but maybe it's easiest to keep that in the edit menu. Up to you 🙂 

Yes, I also felt that pain, but decided against it for now. The thinking here was that if you remove one tag you likely have a better suggestion in mind, in which case you'd go to the edit menu anyway. Again, let's put this to the test and adjust if the driving assumption is too off.

**Comment by mlbrgl** [comment]
### Merge Activity

* **Sep 15, 11:29 AM**: @mlbrgl started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/owid/owid-grapher/2627).

**Pull request merged**